pkg/local-pv-monitor/monitor.go

```diff
@@ -0,0 +1,508 @@
+	newVolume, ok := newObj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", newObj)
+		return
+	}
+
+	monitor.localVolumeMap.UpdateLocalVolume(newVolume)
+}
+
+func (monitor *LocalPVMonitor) deleteVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.DeleteLocalVolume(volume)
+
+}
+
+// Run starts all of this controller's control loops
+func (monitor *LocalPVMonitor) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting monitor controller %s!", string(monitor.RuntimeConfig.Name))
+	monitor.hasRunLock.Lock()
+	monitor.hasRun = true
+	monitor.hasRunLock.Unlock()
+	go monitor.volumeController.Run(stopCh)
+
+	go monitor.MonitorLocalVolumes()
+	<-stopCh
+}
+
+// HasRun returns whether the volume controller has Run
+func (monitor *LocalPVMonitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+// MonitorLocalVolumes checks local PVs periodically
+func (monitor *LocalPVMonitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			pvs := monitor.localVolumeMap.GetPVs()
+			for _, pv := range pvs {
+				monitor.checkStatus(pv)
+			}
+		}
+
+		time.Sleep(DefaultMonitorResyncPeriod)
+	}
+}
+
+// checkStatus checks pv health condition
+func (monitor *LocalPVMonitor) checkStatus(pv *v1.PersistentVolume) {
+	// check if PV is local storage
+	if pv.Spec.Local == nil {
+		glog.Infof("PV: %s is not local storage", pv.Name)
+		return
+	}
+	// check node and pv affinity
+	fit, err := CheckNodeAffinity(pv, monitor.Node.Labels)
+	if err != nil {
+		glog.Errorf("check node affinity error: %v", err)
+		return
+	}
+	if !fit {
+		glog.Errorf("pv: %s does not belong to this node: %s", pv.Name, monitor.Node.Name)
+		return
+	}
+
+	// check if host dir still exists
+	mountPath, continueThisCheck := monitor.checkHostDir(pv)
+	if !continueThisCheck {
+		glog.Errorf("Host dir is modified, PV should be marked")
+		return
+	}
+
+	// check if it is still a mount point
+	continueThisCheck = monitor.checkMountPoint(mountPath, pv)
+	if !continueThisCheck {
+		glog.Errorf("Retrieving mount points error or %s is not a mount point any more", mountPath)
+		return
+	}
+
+	// check PV size: PV capacity must not be greater than device capacity and PV used bytes must not be greater that PV capacity
+	dir, _ := monitor.VolUtil.IsDir(mountPath)
+	if dir {
+		monitor.checkPVAndFSSize(mountPath, pv)
+	}
+	bl, _ := monitor.VolUtil.IsBlock(mountPath)
+	if bl {
+		monitor.checkPVAndBlockSize(mountPath, pv)
+	}
+}
+
+func (monitor *LocalPVMonitor) checkMountPoint(mountPath string, pv *v1.PersistentVolume) bool {
+	// Retrieve list of mount points to iterate through discovered paths (aka files) below
+	mountPoints, mountPointsErr := monitor.RuntimeConfig.Mounter.List()
```

Comment:
this operation can be potentially heavy if there's a lot of pods (list mount point), any thoughts on optimizing this?

Reply:
cache mountpoints ? is this needed ? we just need to list mountpoints on this node. And each node will have a daemonset to do this.